src/java/org/apache/cassandra/index/sai/disk/v1/TermsReader.java

```diff
@@ -203,6 +214,93 @@ public PostingsReader getPostingReader(long offset) throws IOException
         }
     }
 
+    public class RangeQuery
+    {
+        private final QueryEventListener.TrieIndexEventListener listener;
+        private final long lookupStartTime;
+        private final QueryContext context;
+
+        private final Expression exp;
+
+        RangeQuery(Expression exp, QueryEventListener.TrieIndexEventListener listener, QueryContext context)
+        {
+            this.listener = listener;
+            this.exp = exp;
+            lookupStartTime = System.nanoTime();
+            this.context = context;
+        }
+
+        public PostingList execute()
+        {
+            final ByteComparable lower = exp.lower != null ? ByteComparable.fixedLength(exp.getLowerBound()) : null;
+            final ByteComparable upper = exp.upper != null ? ByteComparable.fixedLength(exp.getUpperBound()) : null;
+            try (TrieRangeIterator reader = new TrieRangeIterator(termDictionaryFile.instantiateRebufferer(),
+                                                                  termDictionaryRoot,
+                                                                  lower,
+                                                                  upper,
+                                                                  exp.lowerInclusive,
+                                                                  exp.upperInclusive))
+            {
+                var iter = reader.iterator();
+                if (!iter.hasNext())
+                    return PostingList.EMPTY;
+
+                context.checkpoint();
+                // Because postings are not sorted, we need to eagerly materialize the results and sort them.
+                LongHeap postings = materializeResults(iter);
+
+                listener.onTraversalComplete(System.nanoTime() - lookupStartTime, TimeUnit.NANOSECONDS);
+
+                return new LongHeapPostingList(postings);
+            }
+            catch (Throwable e)
+            {
+                if (!(e instanceof AbortedOperationException))
+                    logger.error(indexContext.logMessage("Failed to execute term query"), e);
+
+                throw Throwables.cleaned(e);
+            }
+        }
+
+        /**
+         * Build an in-memory heap of row ids from the posting lists of the matching terms.
+         * @return an ordered {@link LongHeap} of row ids
+         */
+        private LongHeap materializeResults(Iterator<Pair<ByteSource,Long>> trieEntries) throws IOException
```

Comment:
I have a concern about heap pressure here if we had a very large index with a very low cardinality map, such that we end up with a large number of postings in memory.

Reply:
I think this could be an issue. The issue is the requirement that the PKs produced by the search must be in PK order for intersections to work correctly. Given that we store the data in the trie in term order, I don't see a way around this at the index level.

It would be a hack, but one option could be to have special logic for intersections so that in those queries, we defer validating the predicate until we actually read the row from storage.

Reply:
It's more likely that we should have a memory limiter on these heaps and use a guardrail to fail queries that try to exceed memory limitations. It's nasty but will keep the server from OOMing.

Reply:
How does the memory limiter work on the replica side? I've seen replica failures result in ambiguous errors on the coordinator.